File: brain/src/Application/Common/Models/Contracts.cs
namespace Brain.Application.Common.Models;

public sealed record TimeframeDataContract(
    string Timeframe,
    decimal Open,
    decimal High,
    decimal Low,
    decimal Close,
    long Volume = 0L,
    DateTimeOffset? CandleStartTime = null,
    DateTimeOffset? CandleCloseTime = null,
    decimal CandleBodySize = 0m,
    decimal UpperWickSize = 0m,
    decimal LowerWickSize = 0m,
    decimal CandleRange = 0m,
    decimal Ma20Value = 0m,
    decimal Ma20Distance = 0m,
    decimal Rsi = 0m,
    decimal Atr = 0m);

public sealed record MarketSnapshotContract(
    string Symbol,
    IReadOnlyCollection<TimeframeDataContract> TimeframeData,
    decimal Atr,
    decimal Adr,
    decimal Ma20,
    string Session,
    DateTimeOffset Timestamp,
    decimal Ma20H4 = 0m,
    decimal Ma20H1 = 0m,
    decimal Ma20M30 = 0m,
    decimal RsiH1 = 0m,
    decimal RsiM15 = 0m,
    decimal AtrH1 = 0m,
    decimal AtrM15 = 0m,
    decimal PreviousDayHigh = 0m,
    decimal PreviousDayLow = 0m,
    decimal WeeklyHigh = 0m,
    decimal WeeklyLow = 0m,
    decimal DayOpen = 0m,
    decimal WeekOpen = 0m,
    decimal SessionHigh = 0m,
    decimal SessionLow = 0m,
    decimal SessionHighJapan = 0m,
    decimal SessionLowJapan = 0m,
    decimal SessionHighIndia = 0m,
    decimal SessionLowIndia = 0m,
    decimal SessionHighLondon = 0m,
    decimal SessionLowLondon = 0m,
    decimal SessionHighNy = 0m,
    decimal SessionLowNy = 0m,
    decimal PreviousSessionHigh = 0m,
    decimal PreviousSessionLow = 0m,
    decimal Ema50H1 = 0m,
    decimal Ema200H1 = 0m,
    decimal AdrUsedPct = 0m,
    decimal VolatilityExpansion = 0m,
    DayOfWeek DayOfWeek = default,
    DateTimeOffset Mt5ServerTime = default,
    DateTimeOffset KsaTime = default,
    DateTimeOffset UaeTime = default,
    DateTimeOffset IndiaTime = default,
    DateTimeOffset InternalClockUtc = default,
    DateTimeOffset UtcReferenceTime = default,
    decimal TimeSkewMs = 0m,
    int Mt5ToKsaOffsetMinutes = 50,
    string TelegramImpactTag = "LOW",
    string TradingViewConfirmation = "NEUTRAL",
    bool IsCompression = false,
    bool IsExpansion = false,
    bool IsAtrExpanding = false,
    bool HasOverlapCandles = false,
    bool HasImpulseCandles = false,
    bool HasLiquiditySweep = false,
    bool HasPanicDropSequence = false,
    bool IsPostSpikePullback = false,
    bool IsLondonNyOverlap = false,
    bool IsBreakoutConfirmed = false,
    bool IsUsRiskWindow = false,
    bool IsFriday = false,
    decimal Bid = 0m,
    decimal Ask = 0m,
    decimal Spread = 0m,
    decimal SpreadMedian60m = 0m,
    decimal SpreadMax60m = 0m,
    int CompressionCountM15 = 0,
    int CompressionCountM5 = 0,
    int ExpansionCountM15 = 0,
    decimal ImpulseStrengthScore = 0m,
    string TelegramState = "QUIET",
    bool PanicSuspected = false,
    string TvAlertType = "NONE",
    string SessionPhase = "UNKNOWN",
    // Spread stats per 1m and 5m windows (spec_v5.md A1)
    decimal SpreadMin1m = 0m,
    decimal SpreadAvg1m = 0m,
    decimal SpreadMax1m = 0m,
    decimal SpreadMin5m = 0m,
    decimal SpreadAvg5m = 0m,
    decimal SpreadMax5m = 0m,
    // Account state for exposure cap enforcement (spec_v5.md A1)
    decimal FreeMargin = 0m,
    decimal Equity = 0m,
    decimal Balance = 0m,
    // Tick/market quality (PRD)
    decimal TickRatePer30s = 0m,
    bool FreezeGapDetected = false,
    decimal SlippageEstimatePoints = 0m,
    decimal SessionVwap = 0m,
    decimal SystemFetchedGoldRate = 0m,
    decimal RateDeltaUsd = 0m,
    string RateAuthority = "MT5",
    decimal AuthoritativeRate = 0m,
    string? CycleId = null,
    // Compression metric support (PRD)
    IReadOnlyCollection<decimal>? CompressionRangesM15 = null,
    // Pending/open/execution snapshots (PRD)
    IReadOnlyCollection<PendingOrderSnapshotContract>? PendingOrders = null,
    IReadOnlyCollection<OpenPositionSnapshotContract>? OpenPositions = null,
    IReadOnlyCollection<OrderExecutionEventContract>? OrderExecutionEvents = null);
    
public sealed record PendingOrderSnapshotContract(
    string Type,
    decimal Price,
    decimal Tp,
    DateTimeOffset? Expiry,
    decimal VolumeGramsEquivalent = 0m);

public sealed record OpenPositionSnapshotContract(
    decimal EntryPrice,
    decimal CurrentPnlPoints,
    decimal Tp,
    decimal VolumeGramsEquivalent = 0m);

public sealed record OrderExecutionEventContract(
    string Status,
    DateTimeOffset Timestamp,
    decimal Price = 0m,
    decimal VolumeGramsEquivalent = 0m,
    ulong Ticket = 0UL);

public sealed record TradeSignalContract(
    string Rail,
    decimal Entry,
    decimal Tp,
    DateTimeOffset Pe,
    int Ml,
    decimal Confidence,
    string SafetyTag = "CAUTION",
    string DirectionBias = "BULLISH",
    decimal AlignmentScore = 0.5m,
    string NewsImpactTag = "LOW",
    string TvConfirmationTag = "NEUTRAL",
    IReadOnlyCollection<string>? NewsTags = null,
    string Summary = "",
    bool ConsensusPassed = true,
    int AgreementCount = 1,
    int RequiredAgreement = 1,
    string? DisagreementReason = null,
    IReadOnlyCollection<string>? ProviderVotes = null,
    string ModeHint = "UNKNOWN",
    decimal ModeConfidence = 0.5m,
    int ModeTtlSeconds = 900,
    IReadOnlyCollection<string>? ModeKeywords = null,
    string RegimeTag = "STANDARD",
    string RiskState = "CAUTION",
    string GeoHeadline = "NONE",
    string DxyBias = "NEUTRAL",
    string YieldsBias = "NEUTRAL",
    string CrossMetalsBias = "NEUTRAL",
    string CbFlow = "UNKNOWN",
    string InstPositioning = "UNKNOWN",
    string EventRisk = "LOW",
    IReadOnlyCollection<string>? PromptRefs = null,
    IReadOnlyCollection<string>? ProviderModels = null,
    string? AiTraceJson = null,
    string? CycleId = null);

public sealed record ModeSignalContract(
    string Mode,
    decimal Confidence,
    IReadOnlyCollection<string> Keywords,
    int TtlSeconds,
    DateTimeOffset CapturedAtUtc);

public sealed record RegimeClassificationContract(
    string Regime,
    string RiskTag,
    bool IsBlocked,
    bool IsWaterfall,
    string Reason);

public sealed record DecisionResultContract(
    bool IsTradeAllowed,
    string Status,
    string EngineState,
    string Mode,
    string Cause,
    string WaterfallRisk,
    string Reason,
    string Bucket,
    string Rail,
    string Session,
    string SessionPhase,
    string RegimeTag,
    string RiskState,
    string SizeClass,
    decimal Entry,
    decimal Tp,
    decimal Grams,
    DateTimeOffset ExpiryUtc,
    int MaxLifeSeconds,
    decimal AlignmentScore,
    string TelegramState,
    string RailPermissionA,
    string RailPermissionB,
    int RotationCapThisSession,
    // Section 8.2: Full TABLE columns — shop prices and dual-timezone expiry
    decimal ShopBuy = 0m,
    decimal ShopSell = 0m,
    DateTimeOffset ExpiryKSA = default,
    DateTimeOffset ExpiryServer = default);

public sealed record LedgerStateContract(
    decimal CashAed,
    decimal GoldGrams,
    decimal OpenExposurePercent,
    decimal DeployableCashAed,
    int OpenBuyCount,
    decimal GoldAedEquivalent = 0m,
    decimal NetEquityAed = 0m,
    decimal PurchasePowerAed = 0m,
    decimal DeployedAed = 0m,
    decimal OpenPositionsAed = 0m,
    decimal PendingReservedAed = 0m,
    decimal StartingInvestmentAed = 0m,
    decimal EquityMultiple = 0m,
    // Section 4: C1/C2 bucket capacity (C1=80%, C2=20% of deployable cash)
    decimal BucketC1Aed = 0m,
    decimal BucketC2Aed = 0m);

public sealed record TradeSlipContract(
    string SlipType,
    Guid TradeId,
    decimal Grams,
    decimal Mt5Price,
    decimal ShopPrice,
    decimal AmountAed,
    decimal NetProfitAed,
    decimal CashBalanceAed,
    decimal GoldBalanceGrams,
    DateTimeOffset Mt5Time,
    DateTimeOffset KsaTime,
    string Message);

public sealed record TradingViewSignalContract(
    string Symbol,
    string Timeframe,
    string Signal,
    string ConfirmationTag,
    string Bias,
    string RiskTag,
    decimal Score,
    decimal Volatility,
    DateTimeOffset Timestamp,
    string Source,
    string Notes);

public sealed record TradeCommandContract(
    string Type,
    decimal Price,
    decimal Tp,
    DateTimeOffset Expiry,
    int Ml);

public sealed record PendingTradeContract(
    Guid Id,
    string Symbol,
    string Type,
    decimal Price,
    decimal Tp,
    DateTimeOffset Expiry,
    int Ml,
    decimal Grams = 0m,
    decimal AlignmentScore = 0m,
    string Regime = "",
    string RiskTag = "",
    string EngineState = "ARMED",
    string Mode = "EXHAUSTION",
    string Cause = "UNKNOWN",
    string WaterfallRisk = "LOW",
    string Bucket = "C1",
    string Session = "",
    string SessionPhase = "UNKNOWN",
    string RegimeTag = "STANDARD",
    string RiskState = "CAUTION",
    string SizeClass = "25%",
    string TelegramState = "QUIET",
    bool ConsensusPassed = true,
    int AgreementCount = 1,
    int RequiredAgreement = 1,
    string? DisagreementReason = null,
    IReadOnlyCollection<string>? ProviderVotes = null,
    string Summary = "",
    string ModeHint = "UNKNOWN",
    decimal ModeConfidence = 0.5m,
    string? CycleId = null,
    // Section 8.2: Full TABLE columns — shop prices and dual-timezone expiry
    decimal ShopBuy = 0m,
    decimal ShopSell = 0m,
    DateTimeOffset ExpiryKSA = default,
    DateTimeOffset ExpiryServer = default);

/// <summary>
/// Score breakdown produced by TradeScoreCalculator for a valid trade setup.
/// Score range: 0–100. Thresholds: &lt;45 NO_TRADE, 45–59 WEAK, 60–79 VALID, ≥80 HIGH_CONVICTION.
/// </summary>
public sealed record TradeScoreContract(
    int StructureScore,
    int MomentumScore,
    int ExecutionScore,
    int AiScore,
    int SentimentScore,
    int TotalScore,
    string DecisionTier);

/// <summary>
/// Context passed to the aiworker /study-analyze endpoint when STUDY_LOCK is active.
/// Contains information about recent waterfall failures and blocked trade candidates.
/// </summary>
public sealed record StudyContextContract(
    int ConsecutiveWaterfallFailures,
    string StudyCycleId,
    IReadOnlyCollection<object> RecentBlockedCandidates,
    IReadOnlyCollection<string> RecentWaterfallReasons);

/// <summary>
/// Result from the aiworker autonomous study/self-crosscheck refinement loop.
/// Verdicts: bottomPermissionVerdict = TOO_STRICT | CORRECT | TOO_LOOSE,
///           waterfallVerdict = CORRECT | OVER_SENSITIVE | UNDER_SENSITIVE.
/// </summary>
public sealed record StudyRefinementSuggestionContract(
    string StudyCycleId,
    string BottomPermissionVerdict,
    string WaterfallVerdict,
    IReadOnlyCollection<string> RuleAdjustments,
    double Confidence,
    string Reasoning,
    IReadOnlyCollection<string> ProviderVotes);

File: aiworker/app/models/contracts.py
from datetime import datetime
from typing import Any
from pydantic import BaseModel, Field


class TimeframeData(BaseModel):
    timeframe: str
    open: float
    high: float
    low: float
    close: float
    volume: int = 0
    candleStartTime: datetime | None = None
    candleCloseTime: datetime | None = None
    candleBodySize: float = 0.0
    upperWickSize: float = 0.0
    lowerWickSize: float = 0.0
    candleRange: float = 0.0
    ma20Value: float = 0.0
    ma20Distance: float = 0.0
    rsi: float = 0.0
    atr: float = 0.0


class MarketSnapshot(BaseModel):
    symbol: str = Field(min_length=1, max_length=20)
    timeframeData: list[TimeframeData]
    atr: float
    adr: float
    ma20: float
    ma20H4: float = 0.0
    ma20H1: float = 0.0
    ma20M30: float = 0.0
    rsiH1: float = 0.0
    rsiM15: float = 0.0
    atrH1: float = 0.0
    atrM15: float = 0.0
    previousDayHigh: float = 0.0
    previousDayLow: float = 0.0
    weeklyHigh: float = 0.0
    weeklyLow: float = 0.0
    dayOpen: float = 0.0
    weekOpen: float = 0.0
    sessionHigh: float = 0.0
    sessionLow: float = 0.0
    sessionHighJapan: float = 0.0
    sessionLowJapan: float = 0.0
    sessionHighIndia: float = 0.0
    sessionLowIndia: float = 0.0
    sessionHighLondon: float = 0.0
    sessionLowLondon: float = 0.0
    sessionHighNy: float = 0.0
    sessionLowNy: float = 0.0
    previousSessionHigh: float = 0.0
    previousSessionLow: float = 0.0
    ema50H1: float = 0.0
    ema200H1: float = 0.0
    adrUsedPct: float = 0.0
    session: str
    timestamp: datetime
    volatilityExpansion: float | None = None
    dayOfWeek: str | None = None
    mt5ServerTime: datetime | None = None
    ksaTime: datetime | None = None
    uaeTime: datetime | None = None
    indiaTime: datetime | None = None
    internalClockUtc: datetime | None = None
    utcReferenceTime: datetime | None = None
    timeSkewMs: float = 0.0
    mt5ToKsaOffsetMinutes: int = 50
    telegramImpactTag: str = "LOW"
    tradingViewConfirmation: str = "NEUTRAL"
    isCompression: bool = False
    isExpansion: bool = False
    isAtrExpanding: bool = False
    hasOverlapCandles: bool = False
    hasImpulseCandles: bool = False
    hasLiquiditySweep: bool = False
    hasPanicDropSequence: bool = False
    isPostSpikePullback: bool = False
    isLondonNyOverlap: bool = False
    isBreakoutConfirmed: bool = False
    isUsRiskWindow: bool = False
    isFriday: bool = False
    bid: float = 0.0
    ask: float = 0.0
    spread: float = 0.0
    spreadMedian60m: float = 0.0
    spreadMax60m: float = 0.0
    # Spread stats 1m/5m (spec_v5.md A1)
    spreadMin1m: float = 0.0
    spreadAvg1m: float = 0.0
    spreadMax1m: float = 0.0
    spreadMin5m: float = 0.0
    spreadAvg5m: float = 0.0
    spreadMax5m: float = 0.0
    compressionCountM15: int = 0
    expansionCountM15: int = 0
    impulseStrengthScore: float = 0.0
    telegramState: str = "QUIET"
    panicSuspected: bool = False
    tvAlertType: str = "NONE"
    sessionPhase: str = "UNKNOWN"
    # Account state (spec_v5.md A1)
    freeMargin: float = 0.0
    equity: float = 0.0
    balance: float = 0.0
    # Tick/market quality (PRD)
    tickRatePer30s: float = 0.0
    freezeGapDetected: bool = False
    slippageEstimatePoints: float = 0.0
    sessionVwap: float = 0.0
    systemFetchedGoldRate: float = 0.0
    rateDeltaUsd: float = 0.0
    rateAuthority: str = "MT5"
    authoritativeRate: float = 0.0
    cycleId: str | None = None
    # Compression and order/account snapshots (PRD)
    compressionRangesM15: list[float] = Field(default_factory=list)
    pendingOrders: list[dict[str, Any]] = Field(default_factory=list)
    openPositions: list[dict[str, Any]] = Field(default_factory=list)
    orderExecutionEvents: list[dict[str, Any]] = Field(default_factory=list)


class PostTradeAnalysisRequest(BaseModel):
    """Request for post-trade re-analysis after order placement (PRD point 7)."""
    tradeId: str
    placedRail: str
    placedEntry: float
    placedTp: float
    placedGrams: float
    snapshot: MarketSnapshot


class PostTradeAnalysisSuggestion(BaseModel):
    """Suggestion returned after post-trade re-analysis."""
    tradeId: str
    action: str  # KEEP | ADJUST_ENTRY | ADJUST_TP | CANCEL
    suggestedEntry: float | None = None
    suggestedTp: float | None = None
    confidence: float
    reasoning: str


class TradeSignal(BaseModel):
    rail: str
    entry: float
    tp: float
    pe: datetime
    ml: int
    confidence: float
    safetyTag: str = "CAUTION"
    directionBias: str = "BULLISH"
    alignmentScore: float = 0.5
    newsImpactTag: str = "LOW"
    tvConfirmationTag: str = "NEUTRAL"
    newsTags: list[str] = Field(default_factory=list)
    summary: str = ""
    consensusPassed: bool = True
    agreementCount: int = 1
    requiredAgreement: int = 1
    disagreementReason: str | None = None
    providerVotes: list[str] = Field(default_factory=list)
    modeHint: str = "UNKNOWN"
    modeConfidence: float = 0.5
    modeTtlSeconds: int = 900
    modeKeywords: list[str] = Field(default_factory=list)
    regimeTag: str = "STANDARD"
    riskState: str = "CAUTION"
    geoHeadline: str = "NONE"
    dxyBias: str = "NEUTRAL"
    yieldsBias: str = "NEUTRAL"
    crossMetalsBias: str = "NEUTRAL"
    cbFlow: str = "UNKNOWN"
    instPositioning: str = "UNKNOWN"
    eventRisk: str = "LOW"
    promptRefs: list[str] = Field(default_factory=list)
    providerModels: list[str] = Field(default_factory=list)
    aiTraceJson: str | None = None
    cycleId: str | None = None
